File: lib/quicksort.rb
def quicksort(arr, start = 0, len = arr.length - start)
  return arr[start..len + start] if len < 2
  raise "start index must be in array" unless start.between?(0, (arr.length - 1))
  raise "range exceeds length of array!" if start + len > arr.length

  pivot_idx = partition!(arr, start, len)
  quicksort(arr, start, pivot_idx - start)
  quicksort(arr, pivot_idx + 1, len - (pivot_idx + 1))

  return arr
end

def partition!(arr, start, len)
  end_idx = start + len - 1
  p_idx = start
  start_idx = start + 1
  while start_idx <= end_idx
	  if arr[start_idx] < arr[p_idx]
      arr[p_idx + 1], arr[start_idx] = arr[start_idx], arr[p_idx + 1]
      arr[p_idx], arr[p_idx + 1] = arr[p_idx + 1], arr[p_idx]
      p_idx += 1
     end
    start_idx += 1
  end

  return p_idx
end

File: practice/quicksort.js
Array.prototype.quicksort = function (start, len) {
  if (len <= 1) return this;

  var length = len || this.length;
  var startIdx = start || 0;
  var pivotIdx = this.partition(startIdx, len);

  this.quicksort(startIdx, pivotIdx - startIdx);
  this.quicksort(pivotIdx + 1, len - (pivotIdx + 1));

  return this;
};

Array.prototype.partition = function (start, len) {
  var endIdx = len + startIdx - 1;
  if (endIdx >= this.length) throw new Error("Index out of bounds");
  var pivotIdx = startIdx;
  while (++startIdx <= endIdx) {
    if (this[startIdx] < this[pivotIdx]) {
      var tmp = this[pivotIdx + 1];
      this[pivotIdx + 1] = this[startIdx];
      this[startIdx] = tmp;
      var pivot = this[pivotIdx];
      this[pivotIdx] = this[pivotIdx + 1];
      this[pivotIdx + 1] = pivot;
      pivotIdx++;
    }
  }
  return pivotIdx;
};
